test/lib/mutation_source_test.cc

```diff
@@ -2332,3 +2332,43 @@ void for_each_schema_change(std::function<void(schema_ptr, const std::vector<mut
     }
     test_mutated_schemas();
 }
+
+class mfo_printer {
+    const schema& _s;
+    const mutation_fragment_opt& _mfo;
+public:
+    mfo_printer(const schema& s, const mutation_fragment_opt& mf) : _s(s), _mfo(mf) { }
+    mfo_printer(const mfo_printer&) = delete;
+    mfo_printer(mfo_printer&&) = delete;
+
+    friend std::ostream& operator<<(std::ostream& os, const mfo_printer& p) {
+        if (p._mfo) {
+            os << mutation_fragment::printer(p._s, *p._mfo);
+        } else {
+            os << "eos";
+        }
+        return os;
+    }
+};
+
+static void compare_readers(const schema& s, flat_mutation_reader& authority, flat_reader_assertions& tested) {
+    while (auto expected = authority(db::no_timeout).get()) {
+        tested.produces(s, *expected);
+    }
+    tested.produces_end_of_stream();
+}
+
+void compare_readers(const schema& s, flat_mutation_reader authority, flat_mutation_reader tested) {
+    auto assertions = assert_that(std::move(tested));
+    compare_readers(s, authority, assertions);
+}
+
+void compare_readers(const schema& s, flat_mutation_reader authority, flat_mutation_reader tested, const std::vector<position_range>& fwd_ranges) {
+    auto assertions = assert_that(std::move(tested));
+    compare_readers(s, authority, assertions);
+    for (auto& r: fwd_ranges) {
+        authority.fast_forward_to(r, db::no_timeout);
+        assertions.fast_forward_to(r);
+        compare_readers(s, authority, assertions);
+    }
```

Comment:
I don't understand this test. Once you drive the readers to EOS, the following fast-forwarding-loop is essentially NOOP.

Reply:
It checks that `tested` returns the same thing as `authority`. After they both reach EOS, it checks that `tested` keeps returning EOS (by comparing it to `authority`, which I presume in this case will keep returning EOS if it's correct).